pyro/contrib/funsor/enum_messenger.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from collections import OrderedDict
+
+from pyro.poutine import Messenger
+from pyro.poutine.trace_messenger import TraceMessenger
+
+from pyro.contrib.funsor import named, to_funsor, to_data
+
+
+class SimpleEnumMessenger(Messenger):
+
+    def _pyro_sample(self, msg):
+
+        import funsor
+
+        if msg["done"] or msg["is_observed"] or msg.get("expand", False) or \
+                msg["infer"].get("enumerate") != "parallel":
+            return
+
+        msg["infer"]["funsor_fn"] = to_funsor(msg["fn"])
+        size = msg["infer"]["funsor_fn"].inputs["value"].dtype
+        msg["infer"]["funsor_value"] = funsor.Tensor(
+            torch.arange(size),
+            OrderedDict([(msg["name"], size)]),
+            funsor.bint(size)
+        )
+
+        msg["value"] = to_data(msg["infer"]["funsor_value"])
```

Comment:
Calling `to_data` in inference code such as `SimpleEnumMessenger` handles positional dimension allocation automatically using only the independence information in `markov`/`NamedMessenger` contexts and the fact that each sample site has a fresh name.